File: app/src/main/java/com/akrapovic/soundkit/community/domain/DriveModeEngine.kt
package com.akrapovic.soundkit.community.domain

import com.akrapovic.soundkit.community.data.BleRepository
import com.akrapovic.soundkit.community.data.DiagnosticsRepository
import com.akrapovic.soundkit.community.data.RuleExecutionLogStore
import com.akrapovic.soundkit.community.data.SettingsStore
import java.util.Calendar
import java.util.TimeZone
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

@Singleton
class DriveModeEngine @Inject constructor(
    private val bleRepository: BleRepository,
    private val settingsStore: SettingsStore,
    private val executionLog: RuleExecutionLogStore,
    private val diagnosticsRepository: DiagnosticsRepository,
) {
    private val mutex = Mutex()
    private var quietJob: Job? = null
    private var activeSessionId: Long? = null
    private var userAdjustedSessionId: Long? = null

    fun onUserValveAdjustment() {
        userAdjustedSessionId = activeSessionId
        quietJob?.cancel()
        quietJob = null
    }

    fun onDisconnect() {
        quietJob?.cancel()
        quietJob = null
        activeSessionId = null
    }

    suspend fun onConnectReady(sessionId: Long, scope: CoroutineScope) {
        val settings = settingsStore.settings.first()
        mutex.withLock {
            if (activeSessionId == sessionId) return
            activeSessionId = sessionId
            userAdjustedSessionId = null
            quietJob?.cancel()
            quietJob = null
        }

        if (!settings.driveModeEnabled || settings.automationPaused) {
            diagnosticsRepository.debug("Drive mode skipped: disabled or paused")
            return
        }
        if (!isReadyForAutomation()) {
            diagnosticsRepository.debug("Drive mode skipped: not ready")
            return
        }

        val quiet = settings.quietStart
        if (quiet.enabled && isQuietWindowActive(quiet)) {
            applyClose("quiet start")
            val holdMs = quiet.holdClosedMinutes.coerceIn(1, 15) * 60_000L
            quietJob = scope.launch {
                delay(holdMs)
                if (activeSessionId != sessionId || userAdjustedSessionId == sessionId) return@launch
                val latest = settingsStore.settings.first()
                if (!latest.driveModeEnabled || latest.automationPaused) return@launch
                applyPreferredMode(latest.preferredValveMode, "quiet end")
            }
        } else {
            applyPreferredMode(settings.preferredValveMode, "connect")
        }
    }

    private suspend fun applyPreferredMode(mode: PreferredValveMode, reason: String) {
        if (userAdjustedSessionId == activeSessionId) return
        if (!isReadyForAutomation()) return

        val valveState = bleRepository.valveState.value
        val result = when (mode) {
            PreferredValveMode.Open -> {
                if (valveState == ValveState.Open) {
                    logApply(mode, reason, skipped = true)
                    return
                }
                bleRepository.openValve()
            }
            PreferredValveMode.Closed -> {
                if (valveState == ValveState.Closed) {
                    logApply(mode, reason, skipped = true)
                    return
                }
                bleRepository.closeValve()
            }
        }
        logApply(mode, reason, result = result)
    }

    private suspend fun applyClose(reason: String) {
        if (userAdjustedSessionId == activeSessionId) return
        if (!isReadyForAutomation()) return

        val valveState = bleRepository.valveState.value
        if (valveState == ValveState.Closed) {
            logApply(PreferredValveMode.Closed, reason, skipped = true)
            return
        }
        logApply(PreferredValveMode.Closed, reason, result = bleRepository.closeValve())
    }

    private suspend fun logApply(
        mode: PreferredValveMode,
        reason: String,
        result: CommandResult? = null,
        skipped: Boolean = false,
    ) {
        val outcome = when {
            skipped -> RuleExecutionOutcome.Skipped
            result is CommandResult.Success -> RuleExecutionOutcome.Success
            result is CommandResult.Failure -> RuleExecutionOutcome.Failed
            else -> RuleExecutionOutcome.Skipped
        }
        val detail = when (result) {
            is CommandResult.Failure -> result.message
            else -> null
        }
        executionLog.append(
            RuleExecutionEntry(
                timestampMillis = System.currentTimeMillis(),
                ruleName = "Drive mode",
                action = mode.name,
                reason = reason,
                outcome = outcome,
                detail = detail,
            ),
        )
    }

    private fun isReadyForAutomation(): Boolean {
        val connection = bleRepository.connectionState.value
        if (connection !is ConnectionState.Connected) return false
        if (bleRepository.valveState.value == ValveState.Unknown) return false
        if (bleRepository.receiverStatusMessage.value != null) return false
        return true
    }

    private fun isQuietWindowActive(quiet: QuietStartSettings): Boolean {
        val calendar = Calendar.getInstance(TimeZone.getDefault())
        val dayOfWeek = (calendar.get(Calendar.DAY_OF_WEEK) + 5) % 7
        if (dayOfWeek !in quiet.daysOfWeek) return false
        val minuteOfDay = calendar.get(Calendar.HOUR_OF_DAY) * 60 + calendar.get(Calendar.MINUTE)
        return minuteOfDay in quiet.windowStartMinute..quiet.windowEndMinute
    }
}

File: ios/SoundKitCommunity/Domain/DriveModeEngine.swift
import Foundation

@MainActor
final class DriveModeEngine {
    private weak var bleManager: BLEManager?
    private weak var valveControl: ValveControlCoordinator?
    private weak var settingsStore: SettingsStore?
    private weak var diagnostics: DiagnosticsStore?

    private var quietTask: Task<Void, Never>?
    private var activeSessionId: Int = 0
    private var currentSessionId: Int?
    private var userAdjustedSessionId: Int?

    func configure(
        ble: BLEManager,
        valveControl: ValveControlCoordinator,
        settings: SettingsStore,
        diagnostics: DiagnosticsStore
    ) {
        bleManager = ble
        self.valveControl = valveControl
        settingsStore = settings
        self.diagnostics = diagnostics
    }

    func onUserValveAdjustment() {
        userAdjustedSessionId = currentSessionId
        quietTask?.cancel()
        quietTask = nil
    }

    func onDisconnect() {
        quietTask?.cancel()
        quietTask = nil
        currentSessionId = nil
    }

    func onConnectReady(sessionId: Int) {
        guard let settingsStore, let bleManager else { return }
        if currentSessionId == sessionId { return }
        currentSessionId = sessionId
        userAdjustedSessionId = nil
        quietTask?.cancel()
        quietTask = nil

        let settings = settingsStore.settings
        guard settings.driveModeEnabled, !settings.automationPaused else {
            diagnostics?.debug("Drive mode skipped: disabled or paused")
            return
        }
        guard bleManager.isReadyForAutomation else {
            diagnostics?.debug("Drive mode skipped: not ready")
            return
        }

        let quiet = settings.quietStart
        if QuietWindowEvaluator.isActive(quiet) {
            applyClose(reason: "quiet start")
            let holdSeconds = quiet.holdClosedMinutes.clamped(to: 1...15) * 60
            quietTask = Task { [weak self] in
                try? await Task.sleep(nanoseconds: UInt64(holdSeconds) * 1_000_000_000)
                guard !Task.isCancelled else { return }
                await self?.applyPreferredAfterQuiet(sessionId: sessionId)
            }
        } else {
            applyPreferredMode(settings.preferredValveMode, reason: "connect")
        }
    }

    private func applyPreferredAfterQuiet(sessionId: Int) {
        guard currentSessionId == sessionId, userAdjustedSessionId != sessionId else { return }
        guard let settingsStore, let bleManager else { return }
        let settings = settingsStore.settings
        guard settings.driveModeEnabled, !settings.automationPaused, bleManager.isReadyForAutomation else { return }
        applyPreferredMode(settings.preferredValveMode, reason: "quiet end")
    }

    private func applyPreferredMode(_ mode: PreferredValveMode, reason: String) {
        guard userAdjustedSessionId != currentSessionId else { return }
        guard let bleManager, bleManager.isReadyForAutomation else { return }
        switch mode {
        case .open:
            if bleManager.valveState == .open {
                diagnostics?.debug("Drive mode skipped open (\(reason))")
                return
            }
            valveControl?.open()
        case .closed:
            if bleManager.valveState == .closed {
                diagnostics?.debug("Drive mode skipped closed (\(reason))")
                return
            }
            valveControl?.close()
        }
        diagnostics?.info("Drive mode \(mode.rawValue) (\(reason))")
    }

    private func applyClose(reason: String) {
        guard userAdjustedSessionId != currentSessionId else { return }
        guard let bleManager, bleManager.isReadyForAutomation else { return }
        if bleManager.valveState == .closed {
            diagnostics?.debug("Drive mode already closed (\(reason))")
            return
        }
        valveControl?.close()
        diagnostics?.info("Drive mode close (\(reason))")
    }

    func nextSessionId() -> Int {
        activeSessionId += 1
        return activeSessionId
    }
}

private extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        min(max(self, range.lowerBound), range.upperBound)
    }
}
